controllers/azuremachine_controller.go

```diff
@@ -109,215 +111,297 @@ func (r *AzureMachineReconciler) SetupWithManager(mgr ctrl.Manager, options cont
+        machineScope.SetReady()
+    case infrav1.VMStateCreating:
+        machineScope.V(2).Info("VM is creating", "id", machineScope.GetVMID())
+        conditions.MarkFalse(machineScope.AzureMachine, infrav1.VMRunningCondition, infrav1.VMNCreatingReason, clusterv1.ConditionSeverityInfo, "")
+        machineScope.SetNotReady()
+    case infrav1.VMStateUpdating:
+        machineScope.V(2).Info("VM is updating", "id", machineScope.GetVMID())
+        conditions.MarkFalse(machineScope.AzureMachine, infrav1.VMRunningCondition, infrav1.VMNUpdatingReason, clusterv1.ConditionSeverityInfo, "")
+        machineScope.SetNotReady()
+    case infrav1.VMStateDeleting:
+        machineScope.Info("Unexpected VM deletion", "id", machineScope.GetVMID())
+        r.Recorder.Eventf(machineScope.AzureMachine, corev1.EventTypeWarning, "UnexpectedVMDeletion", "Unexpected Azure VM deletion")
+        conditions.MarkFalse(machineScope.AzureMachine, infrav1.VMRunningCondition, infrav1.VMDDeletingReason, clusterv1.ConditionSeverityWarning, "")
+        machineScope.SetNotReady()
+    case infrav1.VMStateFailed:
+        machineScope.Error(errors.New("Failed to create or update VM"), "VM is in failed state", "id", machineScope.GetVMID())
+        r.Recorder.Eventf(machineScope.AzureMachine, corev1.EventTypeWarning, "FailedVMState", "Azure VM is in failed state")
+        machineScope.SetFailureReason(capierrors.UpdateMachineError)
+        machineScope.SetFailureMessage(errors.Errorf("Azure VM state is %s", machineScope.VMState()))
+        conditions.MarkFalse(machineScope.AzureMachine, infrav1.VMRunningCondition, infrav1.VMProvisionFailedReason, clusterv1.ConditionSeverityWarning, "")
+        machineScope.SetNotReady()
+        // If VM failed provisioning, delete it so it can be recreated
+        err := ams.DeleteVM(ctx)
+        if err != nil {
+            return reconcile.Result{}, errors.Wrapf(err, "failed to delete VM in a failed state")
+        }
+        return reconcile.Result{}, errors.Wrapf(err, "VM deleted, retry creating in next reconcile")
+    default:
+        machineScope.V(2).Info("VM state is undefined", "id", machineScope.GetVMID())
+        conditions.MarkUnknown(machineScope.AzureMachine, infrav1.VMRunningCondition, "", "")
+        machineScope.SetNotReady()
+    }
+
+    return reconcile.Result{}, nil
 }
 
-func (r *AzureMachineReconciler) reconcileDelete(ctx context.Context, machineScope *scope.MachineScope, clusterScope *scope.ClusterScope) (_ reconcile.Result, reterr error) {
-	machineScope.Info("Handling deleted AzureMachine")
+func (r *AzureMachineReconciler) reconcileAzureMachineIPAddress(machineScope *scope.MachineScope, ctx context.Context, clusterScope *scope.ClusterScope, azureIPPoolScope *scope.AzureIPPoolScope) (reterr error) {
+    machineScope.Info("Reconciling IP Address")
+    machineIP := ""
+    if machineScope.IsControlPlane() {
+        machineIP = clusterScope.APIServerIP()
+    }
```

Comment:
Can you exit here?  or do you need to reconcile ip's below for control plane.

Reply:
Ye for control plane, we just grab the API server IP from the pool and assign it to the machine and mark the ip as allocated.

Reply:
So cannot exit and have to go in reconcile